### PR TITLE
Step Interaction

### PR DESCRIPTION
**Pull request by cmalangone**
Added the step interaction following data team suggestion.
Upgraded to version 0.4.1

### CODE REVIEW

File: src/main/scala/io/opentargets/etl/backend/Interactions.scala [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+    /** generate the interactions from a common Dataframe schema
+      * If causalInteraction is true -> swap (A, B) and add to the dataframe
+      * @param mappingInfo Dataframe with mapping_id, ensembl_id
+      * @return a DataFrame
+      */
+    def generateInteractions(mappingInfo: DataFrame): DataFrame = {
+
+      val interactions = df
+        .withColumn("intA", col("interactorA.id"))
```
Comment (JarrodBaker) [inline src/main/scala/io/opentargets/etl/backend/Interactions.scala:129]: If these `withColumn` calls are only renaming columns we should use `.select(col(interactorA.id.as("intA"))` as the withColumn call has a risk of a StackOverflowError: https://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html#withColumn(colName:String,col:org.apache.spark.sql.Column):org.apache.spark.sql.DataFrame. 
Another option is to bundle it later in the `selectExpr`

**Review by JarrodBaker — APPROVED** [review]
I've left one comment with an optional optimisation: if it's run successfully on a full data set then it's something that can be done when you have time. 

Otherwise looks very good. I like how you structured the objects and used the apply method to keep it clean. Nice comments too!

File: src/main/scala/io/opentargets/etl/backend/spark/Helpers.scala [hunk 2]
```diff
@@ -133,21 +137,71 @@ object Helpers extends LazyLogging {
...
   }
 
+  /** generate a set of String with the union of Columns.
+    * Eg, myCols =( a,c,d) and allCols(a,c,d,e,f,h)
+    * return (a,c,d,e,f,h)
+    * @param myCols the list of the Columns in a specific Dataframe
+    * @param allCols the list of Columns to match
+    * @return a sparksession object
+    */
+  def columnExpr(myCols: Set[String], allCols: Set[String]) = {
```
Comment (mkarmona) [inline src/main/scala/io/opentargets/etl/backend/spark/Helpers.scala:161]: @cmalangone when you use sets is better to stick to set operations. those are clearer to follow. by example

```scala
  def columnExpr(myCols: Set[String], allCols: Set[String]) = {
    val inter = (allCols intersect myCols).map(col)
    val differ = (allCols diff myCols).map(lit(null).as(_))

    inter union differ
  }
```
Although if you want to stick with your way you can write this cleaner
```scala
  def columnExpr(myCols: Set[String], allCols: Set[String]) = {
    allCols map {
      case x if myCols.contains(x) => col(x)
      case x =>  lit(null).as(x)
    }
  }
```

**Review by mkarmona — CHANGES_REQUESTED** [review]

File: src/main/scala/io/opentargets/etl/backend/spark/Helpers.scala (hunk 2 above)
Comment (cmalangone) [reply src/main/scala/io/opentargets/etl/backend/spark/Helpers.scala:161]: I changed the code as you suggest. (set operation).

Comment (cmalangone) [reply src/main/scala/io/opentargets/etl/backend/spark/Helpers.scala:161]: @mkarmona  and @JarrodBaker I changed the code as both of you suggested

File: src/main/scala/io/opentargets/etl/backend/Interactions.scala (hunk 1 above)
Comment (cmalangone) [reply src/main/scala/io/opentargets/etl/backend/Interactions.scala]: Done

**Review by mkarmona — APPROVED** [review]

**Pull request merged**